Std/Data/BitVec/Basic.lean

```diff
@@ -433,3 +433,21 @@ def signExtend (v : Nat) (x : BitVec w) : BitVec v := .ofInt v x.toInt
 @[simp] theorem add_eq (x y : BitVec w)                   : BitVec.add x y = x + y            := rfl
 @[simp] theorem sub_eq (x y : BitVec w)                   : BitVec.sub x y = x - y            := rfl
 @[simp] theorem mul_eq (x y : BitVec w)                   : BitVec.mul x y = x * y            := rfl
+
+/-- Turn a `Bool` into a bitvector of length `1` -/
+def ofBool : Bool → BitVec 1
+  | false => 0
+  | true  => 1
+
+/-- The empty bitvector -/
+def nil : BitVec 0 :=
+  BitVec.zero 0
+
+/-- Append a single bit to the end of a bitvector, using big endian order (see `append`).
+    That is, the new bit is the least significant bit. -/
+def concat {n} (msbs : BitVec n) (lsb : Bool) : BitVec (n+1) := msbs ++ (ofBool lsb)
+
+/-- Prepend a single bit to the front of a bitvector, using big endian order (see `append`).
+    That is, the new bit is the most significant bit. -/
+def cons {n} (msb : Bool) (lsbs : BitVec n) : BitVec (n+1) :=
+  ((ofBool msb) ++ lsbs).cast (Nat.add_comm ..)
```

Comment:
It's not clear to me that doing the cast is a good idea here

Reply:
Can you elaborate on why you think the cast might not be a good idea?
This way, we can easily reuse the existing implementation of append, and have a clear connection with that function.
Note that this is `BitVec.cast`, not the generic casting operation (which you are right to be wary of), and, in particular, we have 
```
@[simp]
theorem cast_ofNat {n m : Nat} (h : n = m) (x : Nat) :
    cast h (BitVec.ofNat n x) = BitVec.ofNat m x := by
  subst h; rfl
```

So its not too hard to show equivalence to an alternative definition that does not use cast
```lean
theorem cons_alt (msb : Bool) (lsbs : BitVec w) :
    cons msb lsbs = BitVec.ofNat (w + 1) (BitVec.toNat (ofBool msb) <<< w ||| lsbs.toNat) := by
  simp [cons, (· ++ ·), Append.append, BitVec.append]
```

Of course, that also means it's not too hard to pick this alternative definition as the main definition and proof a theorem that relates `cons` to `append` and `cast`.
I somewhat prefer the current definition, but I'm happy to switch if you insist.

Reply:
To be clear, what I'm suggesting as the simpler alternative is
```suggestion
def cons {n} (msb : Bool) (lsbs : BitVec n) : BitVec (1+n) :=
  ofBool msb ++ lsbs
```

Reply:
I would be somewhat wary of having the resulting length be `1+n`. The precedent with, for example, `Vector.cons`/`Vector.concat` and `Fin.cons`/`Fin.snoc` is to have the result of both operations be `n+1`-sized

Reply:
Ah, that's a good argument. Perhaps that's worth putting in a docstring or at least a comment.

Reply:
`concat` sounds like a synonym for `append` or `join`. What about `snoc` since that makes the naming clear?   When vectors are in `std` they could use the same naming convention.

Reply:
`Vector.concat` was chosen to be consistent with `List.concat`, which is part of core.
I like both `concat` and `snoc`, but I'm not sure this is worth the churn of a rename in core, and I do care about being consistent (it's already somewhat annoying that `Fin.snoc` deviates from these other two).